src/linear_algebra/cross_product.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2021 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak, Kyle Miller, Eric Wieser
+-/
+
+import data.matrix.notation
+import linear_algebra.bilinear_map
+import linear_algebra.matrix.determinant
+import algebra.lie.basic
+
+/-!
+# Cross products
+
+This module defines the cross product of vectors in $R^3$ for $R$ a commutative ring.
+
+## Main definitions
+
+* `cross_product` is the cross product of pairs of vectors in $R^3$.
+
+## Main results
+
+* `triple_product_eq_det`
+* `jacobi_identity`
+
+## Notation
+
+The locale `matrix` gives the following notation:
+
+* `×₃` for the cross product
+
+## Tags
+
+crossproduct
+-/
+
+open matrix
+open_locale matrix
+
+variables {R : Type*} [comm_ring R]
+
+/-- The cross product of two vectors in $R^3$ for $R$ a commutative ring. -/
+def cross_product : (fin 3 → R) →ₗ[R] (fin 3 → R) →ₗ[R] (fin 3 → R) :=
+begin
+  apply linear_map.mk₂ R (λ (a b : fin 3 → R),
+    ![a 1 * b 2 - a 2 * b 1,
+      a 2 * b 0 - a 0 * b 2,
+      a 0 * b 1 - a 1 * b 0]);
+  intros;
+  simp only [@vec3_add R _ _ _ _ _ _ _,
+    pi.add_apply, algebra.id.smul_eq_mul, matrix.smul_cons, matrix.smul_empty, pi.smul_apply];
+  ring_nf,
+end
```

Comment:
On my machine, this proofs takes 24.6s to elaborate.

Here's a way to take less time (8.52s on my machine).
```lean
lemma vec3_congr {α : Type*} {a₁ a₂ a₃ b₁ b₂ b₃ : α}
  (h₁ : a₁ = b₁) (h₂ : a₂ = b₂) (h₃ : a₃ = b₃) :
  ![a₁, a₂, a₃] = ![b₁, b₂, b₃] :=
by subst_vars

def cross_product : (fin 3 → R) →ₗ[R] (fin 3 → R) →ₗ[R] (fin 3 → R) :=
begin
  apply linear_map.mk₂ R (λ (a b : fin 3 → R),
    ![a 1 * b 2 - a 2 * b 1,
      a 2 * b 0 - a 0 * b 2,
      a 0 * b 1 - a 1 * b 0]);
  intros;
  simp only [@vec3_add R _ _ _ _ _ _ _,
    pi.add_apply, algebra.id.smul_eq_mul, matrix.smul_cons, matrix.smul_empty, pi.smul_apply];
  apply vec3_congr;
  simp [add_comm, add_assoc, add_left_comm, mul_comm, mul_assoc, mul_left_comm,
    add_mul, mul_add, sub_eq_add_neg],
end
```
You can check this with the command `set_option profiler true`.

Even just inserting `apply vec3_congr;` before `ring_nf` brings it down to 11.9s.

Reply:
If the elaboration time is important, I will split it to 4 proofs and keep only the relevant lemmata for each goal.

Reply:
But for now, unless you tell me otherwise, I will just add the `apply vec3_congr;` as your second suggestion says.

Reply:
Going from 24.5s to 8.5s is significant, and it makes things less annoying when you want to edit this file. (That's half a minute before you get to see anything relevant about the file.)

I suppose splitting off the proofs of multilinearity would speed things up more (because Lean parallelizes elaboration of theorems, but it elaborates definitions one at a time). I'm not sure that's worth doing, though, and Lean 4 I understand will not work that way.

What I would do is use the 8.5s suggestion and add a comment `-- TODO ring_nf is slow here`.

@eric-wieser What do you think?

Reply:
In a PM with @madvorak, I came up with a very different way of defining it that seems like it might be useful general machinery:
```lean
import data.fin.vec_notation
import linear_algebra.bilinear_map
import linear_algebra.pi
import algebra.algebra.bilinear

variables {R M₁ M₂ M₃ : Type*}
variables [comm_semiring R] [add_comm_monoid M₁] [add_comm_monoid M₂] [add_comm_monoid M₃]
variables [module R M₁] [module R M₂] [module R M₃]

/-- A bilinear map into `fin n.succ → M₃` can be built out of a map into `M₃` and a map into
`fin n → M₃` -/
@[simps]
def linear_map.vec_cons₂ {n} (f : M₁ →ₗ[R] M₂ →ₗ[R] M₃) (g : M₁ →ₗ[R] M₂ →ₗ[R] (fin n → M₃)) :
  M₁ →ₗ[R] M₂ →ₗ[R] (fin n.succ → M₃) :=
linear_map.mk₂ R (λ m₁ m₂, matrix.vec_cons (f m₁ m₂) (g m₁ m₂))
  (λ x₁ x₂ y, sorry)
  (λ r x y, sorry)
  (λ x y₁ y₂, sorry)
  (λ r x y, sorry)

/-- The empty bilinear map defeq to `matrix.vec_empty` -/
@[simps]
def linear_map.vec_empty₂ : M₁ →ₗ[R] M₂ →ₗ[R] (fin 0 → M₃) :=
linear_map.mk₂ R (λ m₁ m₂, matrix.vec_empty)
  (λ _ _ _, funext $ fin.elim0)
  (λ _ _ _, funext $ fin.elim0)
  (λ _ _ _, funext $ fin.elim0)
  (λ _ _ _, funext $ fin.elim0)

/-- The map with `proj_mul_proj i j x y = x i * y j` -/
@[simps]
def proj_mul_proj {ι : Type*} (i j : ι) : (ι → R) →ₗ[R] (ι → R) →ₗ[R] R :=
((algebra.lmul R R).to_linear_map.compl₂ (linear_map.proj j : (ι → R) →ₗ[R] R)).comp (linear_map.proj i : (ι → R) →ₗ[R] R)

/-- The cross product of two vectors in $R^3$ for $R$ a commutative ring. -/
def cross_product {R} [comm_ring R] :
  (fin 3 → R) →ₗ[R] (fin 3 → R) →ₗ[R] (fin 3 → R) :=
linear_map.vec_cons₂ (proj_mul_proj 1 2 - proj_mul_proj 2 1) $
linear_map.vec_cons₂ (proj_mul_proj 2 0 - proj_mul_proj 0 2) $
linear_map.vec_cons₂ (proj_mul_proj 0 1 - proj_mul_proj 1 0) linear_map.vec_empty₂

localized "infixl ` ×₃ `: 68 := cross_product" in matrix

lemma cross_product_apply {R} [comm_ring R] (a b : fin 3 → R) :
  a ×₃ b = ![a 1 * b 2 - a 2 * b 1,
             a 2 * b 0 - a 0 * b 2,
             a 0 * b 1 - a 1 * b 0] :=
rfl
```

Reply:
OK

Reply:
@kmill I resolved the issue using your suggestion in the meanwhile. However, I reopened this issue due to the new suggestion.

Reply:
@eric-wieser Developing that API seems like another PR to me -- should we keep this one as-is and give a nicer definition later?

(I feel like these aren't *quite* the right combinators to define multilinear maps. I think it should be closer to what we'd expect from the tensor-hom adjunction, and rather than `proj_mul_proj` it should be some sort of product-like function on the standard dual basis vectors.)

Reply:
(Maybe I'm lacking reading comprehension and that's what you already did.)

Reply:
It looks very complicated to me. I'd rather not do it in this PR.

Reply:
I'll make a PR with some of those primitives in shortly, I agree they don't belong in this PR.